tf_agents/agents/ddpg/critic_rnn_network.py

```diff
@@ -62,136 +72,97 @@ def __init__(self,
-    if len(lstm_size) == 1:
-      cell = tf.keras.layers.LSTMCell(lstm_size[0])
-    else:
-      cell = tf.keras.layers.StackedRNNCells(
-          [tf.keras.layers.LSTMCell(size) for size in lstm_size])
+        dtype=dtype,
+        name='lstm')
 
-    state_spec = tf.nest.map_structure(
-        functools.partial(
-            tensor_spec.TensorSpec, dtype=tf.float32,
-            name='network_state_spec'), list(cell.state_size))
-
-    output_layers = utils.mlp_layers(fc_layer_params=output_fc_layer_params,
-                                     name='output')
-
-    output_layers.append(
+    output_layers = [
         tf.keras.layers.Dense(
             1,
             activation=None,
             kernel_initializer=tf.keras.initializers.RandomUniform(
                 minval=-0.003, maxval=0.003),
-            name='value'))
+            name='value')]
 
     super(CriticRnnNetwork, self).__init__(
         input_tensor_spec=input_tensor_spec,
-        state_spec=state_spec,
+        state_spec=lstm_encoder.state_spec,
         name=name)
 
-    self._observation_layers = observation_layers
-    self._action_layers = action_layers
-    self._joint_layers = joint_layers
-    self._dynamic_unroll = dynamic_unroll_layer.DynamicUnroll(cell)
+    self._obs_encoder = obs_encoder
+    self._lstm_encoder = lstm_encoder
     self._output_layers = output_layers
 
   # TODO(kbanoop): Standardize argument names across different networks.
   def call(self, inputs, step_type, network_state=None):
+    outer_rank = nest_utils.get_outer_rank(inputs, self.input_tensor_spec)
+    batch_squash = utils.BatchSquash(outer_rank)
+
     observation, action = inputs
-    observation_spec, _ = self.input_tensor_spec
-    num_outer_dims = nest_utils.get_outer_rank(observation,
-                                               observation_spec)
-    if num_outer_dims not in (1, 2):
-      raise ValueError(
-          'Input observation must have a batch or batch x time outer shape.')
-
-    has_time_dim = num_outer_dims == 2
-    if not has_time_dim:
-      # Add a time dimension to the inputs.
-      observation = tf.nest.map_structure(lambda t: tf.expand_dims(t, 1),
-                                          observation)
-      action = tf.nest.map_structure(lambda t: tf.expand_dims(t, 1), action)
-      step_type = tf.nest.map_structure(lambda t: tf.expand_dims(t, 1),
-                                        step_type)
-
-    observation = tf.cast(tf.nest.flatten(observation)[0], tf.float32)
-    action = tf.cast(tf.nest.flatten(action)[0], tf.float32)
+    observation, _ = self._obs_encoder(
+        observation,
+        step_type=step_type,
+        network_state=network_state)
 
-    batch_squash = utils.BatchSquash(2)  # Squash B, and T dims.
-    observation = batch_squash.flatten(observation)  # [B, T, ...] -> [BxT, ...]
+    observation = batch_squash.flatten(observation)
+    action = tf.cast(tf.nest.flatten(action)[0], tf.float32)
```

Comment:
should this be casting to a user-defined dtype (or if that's not available, then tf.float32)?